### ISSUE TITLE
Pyright does not recognize that field cannot be None

### ISSUE THREAD

**Issue by oefe** [microsoft/pyright#7706]
**Describe the bug**

In the code below, Pyright does not recognize that field is set to value other than `None` on all code paths.

**Code**

```python
import json
from pathlib import Path
from typing import Any


class Settings:
    _settings: dict[str, Any] | None = None

    @property
    def settings(self) -> dict[str, Any]:
        if self._settings is None:
            if Path("settings.json").exists():
                with Path("settings.json").open("r") as f:
                    self._settings = json.load(f)
            else:
                self._settings = {}
        return self._settings
```

This passed in pyright 1.1.357, but pyright 1.1.358 reports an error:
```console
Expression of type "dict[str, Any] | None" cannot be assigned to return type "dict[str, Any]"
  Type "dict[str, Any] | None" cannot be assigned to type "dict[str, Any]"
    "None" is incompatible with "dict[str, Any]"
```

I didn't see anything in the pyright 1.1.358 changes that would indicate that this is an intentional change.

**VS Code extension or command-line**
Observed in both VS Code and command line, using pyright 1.1.358

**Comment by roganov** [comment]
I believe this behavior is correct because `json.load` returns `Any` -- after all, your `settings.json` file could be `null`. You can use `cast` to circumvent the error.

```
import json
from pathlib import Path
from typing import Any, cast


class Settings:
    _settings: dict[str, Any] | None = None

    @property
    def settings(self) -> dict[str, Any]:
        if self._settings is None:
            if Path("settings.json").exists():
                with Path("settings.json").open("r") as f:
                    self._settings = cast(dict[str, Any], json.load(f))
            else:
                self._settings = {}
        return self._settings
```

**Comment by erictraut** [comment]
Yes, the analysis by @roganov is correct. However, I recommend against using a `cast`. This approach effectively blinds the type checker, so it makes your code more fragile. A better approach is to use an `assert`, which not only documents your assumption for readers of the code, it also tells the the type checker about your assumption, and it validates it at runtime.

```python
                with Path("settings.json").open("r") as f:
                    self._settings = json.load(f)
                    assert self._settings is not None
```

**Comment by oefe** [comment]
I'm not sure I understand this argument. Isn't `Any` supposed to be compatible with any type, and therefore suppress this error?

I also tried to change the assignment to `self._settings = json.load(f) or {}`, so that any `None` result from `load` would be replaced by `{}`, which is compatible with `dict[str, Any]`. Still the same result.

However, if I use an explicit `if` statement instead, the error is suppressed as expected:
```python
                    self._settings = json.load(f)
                    if not self._settings:
                        self._settings = {}
```

**Comment by erictraut** [comment]
The type of the instance variable `self._settings` is declared to be `dict[str, Any] | None`. If you were to assign a value of `None` to this variable, pyright would narrow its type upon assignment to just `None`. If you assign the expression `{}` to it, pyright narrows its type upon assignment to just `dict[str, Any]`. If you assign a value of type `Any` to this variable, pyright cannot narrow its type beyond `dict[str, Any] | None`. The assigned `Any` value prevents pyright from narrowing the type.

If you want the type to be narrowed in this case, you can add an `assert` statement as I showed above.

Not surprisingly, mypy behaves the same way.

```python
class Test:
    _settings: dict[str, Any] | None = None

    def test(self):
        self._settings = None
        reveal_type(self._settings)  # None

        self._settings = {}
        reveal_type(self._settings)  # dict[str, Any]

        with Path("settings.json").open("r") as f:
            self._settings = json.load(f)
        reveal_type(self._settings)  # dict[str, Any] | None

        assert self._settings is not None
        reveal_type(self._settings)  # dict[str, Any]
```

If you'd like a better understanding of type declarations and type narrowing, I recommend reading [this documentation](https://microsoft.github.io/pyright/#/type-concepts) and [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced).

**Comment by oefe** [comment]
> Not surprisingly, mypy behaves the same way.

not quite:
```python
class Test:
    _settings: dict[str, Any] | None = None

    def test(self) -> None:
        with Path("settings.json").open("r") as f:
            # Original problem:
            self._settings = json.load(f)
            reveal_type(self._settings)
            # pr 1.1.357: Any
            # pr 1.1.358: dict[str, Any] | None
            # mypy 1.9.0: dict[str, Any] | None
            # Here, pyright 1.1.357 deviated from mypy, and pyright 1.1.358 fixes this

            # Making sure that the result cannot be None:
            self._settings = json.load(f) or {}
            reveal_type(self._settings)
            # pr 1.1.357: Any | dict[str, Any]
            # pr 1.1.358: dict[str, Any] | None
            # mypy 1.9.0: dict[str, Any] | Any
            # Here, pyright 1.1.357 matched mypy, but pyright 1.1.358 thinks the value could be None
            # I think this is incorrect, there is no path that could result in the value being None
            # The result of pyright 1.1.357/mypy appears to be correct. Any if json.load is truish, dict if it is falsish

            # Using if instead of or (logically equivalent, so I would expect the same result):
            self._settings = json.load(f)
            if not self._settings:
                self._settings = {}
            reveal_type(self._settings)
            # pr 1.1.357:  dict[str, Any] | Any
            # pr 1.1.358: dict[str, Any]
            # mypy 1.9.0: dict[str, Any] | Any
            # Here, pyright 1.1.357 matched mypy, but pyright 1.1.358 narrows the value down to a dict
            # I think this is incorrect, the return type of json.load is Any, and if it is a truish value,
            # that would be the type of self._settings.
            # As above, the result of pyright 1.1.357/mypy appears to be correct. Any if json.load is truish, dict if it is falsish```
